src/transports/imap/imap_transport.php

```diff
@@ -575,6 +614,25 @@ public function authenticate( $user, $password )
         return true;
     }
 
+    /**
+     * Tries to login to the SMTP server with 'AUTH LOGIN'
+     */
+    protected function authLogin($user, $password)
+    {
+        $user = addcslashes($user, '\"');
+        $password = addcslashes($password, '\"');
+        $this->connection->sendData( "{$tag} LOGIN \"{$user}\" \"{$password}\"" );
+    }
+
+    /**
+     * Tries to login to the IMAP server with 'AUTH XOAUTH2'
+     */
+    protected function authXOAuth2($user, $password)
+    {
+        $digest = base64_encode("user={$user}\1auth=Bearer {$password}\1\1");
+        $this->connection->sendData( "AUTH XOAUTH2 {$digest}" );
```

Comment:
I believe the `AUTH` should be prefixed with the `{$tag} ` here as well. I think you can:

1. refactor, so that `authLogin` and `authXOAuth2` return the full auth command to send (and rename them to `getLoginCommand` / `getXOauth2Command`, and move the `sendData` back into the calling function.
2. Move the `$tag = $this->getNextTag();` into the  `authLogin` and `authXOAuth2` method
3. Make sure you pass in `$tag` to both `autoLogin` and `autoXOAuth2`

I think I would pick two, and also rename the methods to `sendLogin` and `sendXOAuth2`, and also mark them as private.